open-apparel-registry/open-apparel-registry
# Implement Source model - Step 2

## Overview

The second step of a three step process to add a Source model between facility lists and list items, allowing for the submission of individual facilities.

Each step in the process is intended to be released separately and maintain backward compatibility.

This second step is the most complex and involves switching over all the code to reference a related `Source` model for retrieving the `Contributor` and `is_active`/`is_private` status of a list. 

Connects #817
Depends on #856

## Notes

This is the second of three pull requests to be deployed separately.

~This PR is targeted at `jcw/source-model-1` because it depends on it.~
(Retargeted after merging and deploying `jcw/source-model-1`)

The third step will cover removing the fields from other models that are now present on the `Source`

## Testing Instructions


* Checkout `develop` and `./scripts/resetdb`
* Checkout `jcw/source-model-1`, run `./scripts/manage migrate` and run `./scripts/manage liststosources`
* Checkout this branch
* Exercise every list/item/facility related feature
  * Log in as `c2@example.com`
  * Search for `regina miracle intimate` and view the details of the matching facility. Verify that the "Other locations," "Also known as," "Other addresses," and "Contributors" sections are populated.
  * Click the "Claim this facility" link and verify that you can successfully submit the claim.
  * In a new private browser window
    * Log in as `c1@example.com`.
    * Browse http://localhost:6543/dashboard/claims/1 and approve the claim
  * Back in the original browser window, refresh the "Regina" facility details page and click "Update facility details"
    * Enter a few field values and save the form
  * Copy the OAR ID
  * Browse http://localhost:8081/api/docs/#!/facilities/facilities_get_facility_history and submit the OAR ID. Verify that the history contains `CREATE`, `ASSOCIATE`, `CLAIM`, and `CLAIM_UPDATE`	events.
  * Browse http://localhost:6543/contribute and upload [other-location-test.csv.zip](https://github.com/open-apparel-registry/open-apparel-registry/files/3692205/other-location-test.csv.zip)
  * Fully process the list. Verify there are no processing errors.
    * `./scripts/manage batch_process --list-id 16 --action parse`
    * `./scripts/manage batch_process --list-id 16 --action geocode`
    * `./scripts/manage batch_process --list-id 16 --action match`
  * Browse http://localhost:6543/lists/16
    * Reject one of the pending matches, accept the other. Verify that the process completes without error.
  * Use `./scripts/manage shell_plus` to verify that a `Source` is created for each `FacilityList` (including the new "16") and that a `Source` is assigned on each `FacilityListItem`
```
>>> list(FacilityList.objects.all().order_by('id').values_list('id', flat=True))
[2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]
>>> list(Source.objects.all().order_by('facility_list').values_list('facility_list', flat=True))
[2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16]
>>> FacilityListItem.objects.filter(source=None).count()
0
```

## Checklist
